src/main/java/com/znsio/teswiz/runner/DeviceSetup.java

```diff
@@ -84,15 +85,64 @@ static void verifyAppExistsAtMentionedPath() {
         LOGGER.info(String.format("Update path to Apk: %s", appPath));
         if(appPath.equals(NOT_SET)) {
             appPath = getAppPathFromCapabilities();
+            String directoryPath = "src" + File.separator + "test" + File.separator + "resources" + File.separator + "sampleApps";
+            String fileName = appPath.split(File.separator)[appPath.split(File.separator).length - 1];
+            String filePath = directoryPath + File.separator + fileName;
+            if((new File(filePath)).exists()){
+                LOGGER.info("File already exist at "+filePath);
+                appPath = filePath;
+            }
+            else if (isURLValid(appPath)) {
+                LOGGER.info("URL present instead of file path");
+                File file = new File(fileName);
+                try {
+                    LOGGER.info("Creating repositories if doesn't exist");
+                    Files.createDirectories(Path.of(directoryPath));
+                    if (!file.exists()) {
+                        LOGGER.info("File doesn't exist, need to download it");
+                        downloadFile(appPath, filePath);
+                    }
+                    appPath = filePath;
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
             Setup.addToConfigs(APP_PATH, appPath);
-            String capabilitiesFileName = Setup.getFromConfigs(CAPS);
-            checkIfAppExistsAtTheMentionedPath(appPath, capabilitiesFileName);
         } else {
             LOGGER.info(String.format("\tUsing AppPath provided as environment variable -  %s",
                                       appPath));
         }
     }
 
+    private static boolean isURLValid(String urlString) {
+        try {
+            new URL(urlString);
+            return true;
+        } catch (MalformedURLException e) {
+            return false;
+        }
+    }
+
+    private static void downloadFile(String url, String filePath) throws IOException {
+        LOGGER.info("Starting to download file");
+        URL fileUrl = new URL(url);
```

Comment:
We are creating new url objects, instead of passing string url we can pass URL url.

Reply:
created new URL in method to handle exception in the called method itself